### PR TITLE
adding by_user permissions support for windows

### PR DESCRIPTION
**Pull request by jeremymv2**
This is a first stab at adding mvp support for checking read/write/execute file permissions, for a specific user.

```
  # verifies if a specific user has access to a file
    describe file("C:/users/vagrant/setup.bat") do
      it { should exist }
      it { should_not be_readable.by_user('vagrant-2012-r2\jdoe') }
      it { should be_readable.by_user('vagrant-2012-r2\vagrant') }
      it { should_not be_executable.by_user('vagrant-2012-r2\jdoe') }
      it { should be_executable.by_user('vagrant-2012-r2\vagrant') }
      it { should_not be_writable.by_user('vagrant-2012-r2\jdoe') }
      it { should be_writable.by_user('vagrant-2012-r2\vagrant') }
      it { should be_readable }
      it { should be_writeable }
      it { should be_executable }
    end
```

The above results in the following when `vagrant-2012-r2\jdoe` was denied access to the file:

screen shot 2016-10-11 at 9 59 33 pm

I believe it addresses #783 

It will fail unit tests since they will have to be refactored.  If this passes muster I'll refactor the tests.

Signed-off-by: Jeremy J. Miller jm@chef.io

### CODE REVIEW

**Comment by jeremymv2** [comment]
@chris-rock @arlimus thoughts?

File: lib/resources/file.rb [hunk 1]
```diff
@@ -20,11 +33,13 @@ class FileResource < Inspec.resource(1) # rubocop:disable Metrics/ClassLength
...
       end
     "
-    include MountParser
 
     attr_reader :file, :mount_options
     def initialize(path)
+      # select permissions style
+      @perms_provider = select_file_perms_style(inspec.os)
       @file = inspec.backend.file(path)
+      return skip_resource 'The `file` resource is not supported on your OS yet.' if @perms_provider.nil?
```
Comment (chris-rock) [inline lib/resources/file.rb]: I think we should not skip the resource if the no perm_provider is available. Most of the file implementation is done in train anyway. I would prefer to move this skip return to the individual functions that cannot be executed aka `readable?` etc

File: lib/resources/file.rb [hunk 2]
```diff
@@ -52,19 +67,19 @@ def contain(*_)
     def readable?(by_usergroup, by_specific_user)
       return false unless exist?
```
Comment (chris-rock) [inline lib/resources/file.rb:68]: This is the place where we should return the skip the resource

File: lib/resources/file.rb [hunk 3]
```diff
@@ -109,16 +124,14 @@ def to_s
 
     private
 
-    def file_permission_granted?(flag, by_usergroup, by_specific_user)
-      unless inspec.os.unix?
-        fail 'Checking file permissions is not supported on your os'
-      end
-
+    def file_permission_granted?(access, by_usergroup, by_specific_user)
+      return nil if @perms_provider.nil?
```
Comment (chris-rock) [inline lib/resources/file.rb]: we should fail here, once we have the skip as the first statement of eg. in `readable?` this method should never be called if perms_provider is not defined.

File: lib/resources/file.rb [hunk 4]
```diff
@@ -129,31 +142,63 @@ def check_file_permission_by_mask(usergroup, flag)
...
+  class UnixFilePermissions < FilePermissions
+    def check_file_permission_by_user(user, access_type, path)
+      flag = case access_type
+             when 'read'
+               'r'
+             when 'write'
+               'w'
+             when 'execute'
+               'x'
+             end
```
Comment (chris-rock) [inline lib/resources/file.rb:177]: we should return an error if the access_type is unknown

**Comment by chris-rock** [comment]
Awesome work @jeremymv2 I had some minor comments. Nothing difficult to improve. On top, I would prefer if we could add some integration tests as well in https://github.com/chef/inspec/blob/master/test/integration/default/file_spec.rb#L173-L200

File: lib/resources/file.rb [hunk 5]
```diff
@@ -51,20 +65,23 @@ def contain(*_)
 
     def readable?(by_usergroup, by_specific_user)
       return false unless exist?
+      return skip_resource '`readable?` is not supported on your OS yet.' if @perms_provider.nil?
```
Comment (chris-rock) [inline lib/resources/file.rb:69]: 👍

**Comment by chris-rock** [comment]
Awesome work @jeremymv2, lets rebase, fix the tests and I'll merge it. Let me know if I can help with fixing the tests.

**Comment by jeremymv2** [comment]
@chris-rock unit and integration tests added to the PR.
I heavily refactored the unit tests - let me know what you think.

**Comment by chris-rock** [comment]
Awesome @jeremymv2

**Pull request merged**